pss/trojan_chunk.go

```diff
@@ -0,0 +1,221 @@
+func checkTargets(targets [][]byte) error {
+	if len(targets) == 0 {
+		return fmt.Errorf("target list cannot be empty")
+	}
+	validLength := len(targets[0]) // take first element as allowed length
+	for i := 1; i < len(targets); i++ {
+		if len(targets[i]) != validLength {
+			return fmt.Errorf("target list cannot have targets of different length")
+		}
+	}
+	return nil
+}
+
+// newTrojanChunkSpan creates a pre-set 8-byte span for a trojan chunk
+func newTrojanChunkSpan() []byte {
+	span := make([]byte, 8)
+	binary.BigEndian.PutUint64(span, chunk.DefaultSize) // TODO: should this be little-endian?
+	return span
+}
+
+// serializeTrojanChunk appends the span, nonce and payload of a trojan message and returns the result
+// this can be used:
+// - to form the payload for a regular chunk
+// - to be used as the input for trojan chunk hash calculation
+func serializeTrojanChunk(span, nonce, payload []byte) ([]byte, error) {
+	h := append(span, nonce...)
+	s := append(h, payload...)
+
+	return s, nil
+}
+
+// findNonce determines the nonce so that when the given trojan chunk fields are hashed, the result will fall in the neighbourhood of the given address
+// this is done by iterating the BMT hash of the serialization of a trojan chunk until the desired nonce is found
+// the matching hash is also returned as the target address
+func (tm *trojanMessage) findNonce(span []byte, targets [][]byte) (nonce, address []byte, err error) {
+	emptyNonce, emptyAddress := []byte{}, []byte{}
+
+	// start out with random nonce
+	nonce = make([]byte, 32)
+	if _, randErr := rand.Read(nonce); randErr != nil {
+		return emptyNonce, emptyAddress, randErr
+	}
+	nonceInt := new(big.Int).SetBytes(nonce)
+	targetsLength := len(targets[0])
+
+	// serialize trojan message
+	m, marshalErr := tm.MarshalBinary()
+	if marshalErr != nil {
+		return emptyNonce, emptyAddress, marshalErr
+	}
+
+	// hash trojan chunk fields with different nonces until an acceptable one is found
+	// TODO: prevent infinite loop
+	for {
+		hash, hashErr := hashTrojanChunk(span, nonce, m)
+		if hashErr != nil {
+			return emptyNonce, emptyAddress, hashErr
+		}
+
+		// take as much of the hash as the targets are long
+		if hashPrefixInTargets(hash[:targetsLength], targets) {
+			// if nonce found, stop loop and return matching hash as address
+			return nonce, hash, nil
+		}
+		// else, add 1 to nonce and try again
+		// TODO: test loop-around
+		nonceInt.Add(nonceInt, big.NewInt(1))
+		nonce = nonceInt.Bytes()
+	}
+}
+
+// hashTrojanChunk serializes trojan chunk fields and hashes them with the trojan hashing func
+// returns the resulting hash or a hashing error if it occurs
+func hashTrojanChunk(span, nonce, payload []byte) ([]byte, error) {
+	s, _ := serializeTrojanChunk(span, nonce, payload) // err always nil here
+	trojanHashingFunc.Reset()                          // TODO: why do we need to do this?
```

Comment:
so i slightly adjusted this, but:
- i don't need to actually call `SetSpanBytes`, do i? since i want to calculate the hash for the _entire_ byte slice received, no?
- i should still actually check for `Write` `err`, no?

Reply:
youo absolutely  do need to call `SetSpanBytes` with the span bytes and `Write` to the. hash only the payload

Reply:
i've been looking at the code and i think i understand why now, thanks.

this also made it possible for the code to work with `BMTHash` :heart: